### ISSUE TITLE
Both `cargo check` and `cargo clippy` one after the other in rust-analyzer

### ISSUE THREAD

**Issue by frederikhors** [rust-lang/rust-analyzer#21286]
Hi everyone,

In my `.vscode/settings.json` I've configured `cargo clippy` as the default value for `rust-analyzer.check.overrideCommand`. While this gives useful linting, it significantly slows down feedback compared to `cargo check`, which is much faster.

Is there a way to combine both workflows? Ideally, I'd like fast feedback from `cargo check` and right after (and maybe only if there are NO errors) `cargo clippy`.

Is this possible with `rust-analyzer` or VS Code, and if so, how can it be configured?

What do you like?

**Comment by ChayimFriedman2** [comment]
AFAIK rust-analyzer supports streaming the errors from Cargo, so you can create a Cargo wrapper that calls them both and writes both results (but you may need coordination to handle them). That's probably a bad idea though.

Other than that, I don't believe r-a is going to officially support running both. Running one is already too slow for many, and has significant complexity in the r-a codebase.

**Comment by ChayimFriedman2** [comment]
Also note that the ideal solution is to make native r-a diagnostics which are instantaneous, usually. There are many of them already (although definitely not exhaustive), most are disabled by default (for having false positives, although for many the false positive rate is low today). You can enable them by setting [`"rust-analyzer.diagnostics.experimental.enable": true`](https://rust-analyzer.github.io/book/configuration.html#diagnostics.experimental.enable).

**Comment by frederikhors** [comment]
Are the r-a catching the errors like `cargo-check` does?

Anyway I'm asking to have a way to launch `cargo check` and only then `cargo clippy` and only if I want. Not by default for everyone.

Can we make this happen with a config?

**Comment by ChayimFriedman2** [comment]
> Are the r-a catching the errors like `cargo-check` does?

Some of them.

> Anyway I'm asking to have a way to launch `cargo check` and only then `cargo clippy` and only if I want. Not by default for everyone.
> 
> Can we make this happen with a config?

As I said:

> Running one is already too slow for many, and has significant complexity in the r-a codebase.

So: too much implementation complexity, too little users who want this. My answer is no, but @rust-lang/rust-analyzer do you agree?

**Comment by frederikhors** [comment]
I mean I need something like:

```json
"rust-analyzer.check.overrideCommand": [
    "cargo check",
    "&&",
    "cargo clippy"
],
```

Maybe this already works! I should try it.

**Comment by frederikhors** [comment]
```log
cargo check failed to start: Cargo watcher failed, the command produced no valid metadata (exit code: ExitStatus(ExitStatus(1))): error: unexpected argument '&&' found
```

**Comment by ChayimFriedman2** [comment]
I understand what you want, and except for the ugly possible workaround that I posted in the beginning, I said that I don't want r-a to support this. But let's wait and see what other team members say.

**Comment by frederikhors** [comment]
Oh, got it - thanks so much for explaining. I think I could have been clearer on my end. I really appreciate all your hard work!

**Comment by ShoyuVanilla** [comment]
I haven't tried this yet by myself but I guess this would run `cargo check` first, and then `cargo clippy`.

```json
"rust-analyzer.check.overrideCommand": "sh -c \"cargo check --workspace --message-format=json --all-targets && cargo clippy --workspace --message-format=json --all-targets\""
```

My take is:

- flycheck diagnostics are applied by stream, so even though you execute both commands, it only slows down the moment when they are completed. You'll get each diagnostics as soon as they are emitted from the commands
- And if you trigger another flycheck by saving a file, the previous job is cancelled, so the increased flycheck duration won't block your workflow, except it is consuming more system resources 😅 
- And we even deduplicate flycheck diagnostics, so okay with having the same diagnostics from the clippy and normal check

But for the first bulletpoint, I think all the above thing might be pointless. We already applies the flycheck diagnostics incrementally. Will running `cargo check` and then `cargo clippy` speed up feedback loop, unless you run them parallel?

**Comment by frederikhors** [comment]
@ShoyuVanilla If I understand what you wrote, even if I use the command with `sh -c`, it doesn't make sense because I'd receive the error stream when both commands are finished.

This leaves me a bit perplexed and sad: that there isn't a way to run one command first and then the other, not in parallel but consecutively.

It would be very useful because there would be a very fast `check` first and only after (if there are no errors with `check`) a much slower `clippy`.

Sorry for the further repetition, but since I'm not an expert in Rust-Analyzer, I don't understand the difficulty and only see the advantages.

**Comment by ShoyuVanilla** [comment]
> I'd receive the error stream when both commands are finished.

No, you receive them as soon as they are emitted from stdout/stderr

**Comment by frederikhors** [comment]
Oh, ok. So you're command can work for me, right? I don't understand why for you:

> But for the first bulletpoint, I think all the above thing might be pointless.

**Comment by ShoyuVanilla** [comment]
I mean, since the diagnostics from clippy are streamed as soon as they are emitted as well, rather than once finished, so there's little reason to run normal check and clippy consecutively for responsiveness

**Comment by frederikhors** [comment]
And that's the whole reason I opened this issue!

If I run `cargo check`, it's VERY FAST! Let's say 3 seconds.

If I run `cargo clippy`, it's VERY SLOW! Let's say 20 seconds.

For the same errors.

**Comment by ShoyuVanilla** [comment]
That's strange. But I guess that might because of clippy just being slow on emitting diagnostics, for its internal lint passes. In that case, running things in serial would make sense.

**Comment by ChayimFriedman2** [comment]
Given no other team member has objected, I'm closing this.
